Applied Connect/Disconnect verbs to Add-/Remove-AzureRmAccount cmdlets

KirkMunro opened the pull request:
## Description

This PR renames two cmdlets in AzureRM.Profile:
- Add-AzureRmAccount has been renamed as Connect-AzureRmAccount
- Remove-AzureRmAccount has been renamed as Disconnect-AzureRmAccount

In addition, all use of Add-AzureRmAccount and Login-AzureRmAccount have been replaced by Connect-AzureRmAccount, and all use of Remove-AzureRmAccount and Logout-AzureRmAccount have been replaced by Disconnect-AzureRmAccount.

The original cmdlets names (Add-/Remove-) have been added as aliases for the new, more accurate and more appropriate command names.

Whenever you find yourself using an alias in place of a cmdlet name in documentation or during presentations, you really should ask yourself: why are you using an alias instead of the cmdlet itself? In this case, the answer is quite simple: Add-/Remove- are not very appropriate verbs for connecting to/disconnecting from a system or service; Connect and Disconnect are the verbs that should be used in such scenarios, and their name is clear enough that there is no need to use an alias to try to clarify what the command is actually doing.

I have a [separate PR for azure-docs](https://github.com/MicrosoftDocs/azure-docs/pull/3658) where I have applied the same broad changes across all open source azure documentation, replacing Add-/Login-AzureRmAccount with Connect-AzureRmAccount everywhere that the aliases/poor command names were used. That PR should be approved once this PR is approved, merged, and released.

Moving forward beyond this PR, use of aliases during demos and in presentations should be discouraged unless the purpose of those demos is to educate users about the aliases. Use of aliases in documentation should be forbidden unless the purpose of that documentation is to educate users about those aliases. In this case, there are many unnecessary aliases for these commands at this point. It would be more beneficial for the community to phase these aliases out of their work completely than to continue using them, and the only way that will happen is with proper leadership demonstrating the right way to use commands that have more appropriate names.

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [X] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [X] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [X] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [X] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [X] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.
- [ ] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [ ] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [ ] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [ ] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [ ] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [ ] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

KirkMunro commented:
Here is the corresponding PR for azure-docs: https://github.com/MicrosoftDocs/azure-docs/pull/3658. That PR should be accepted once this PR is approved and included in a release.

KirkMunro commented:
This is building fine on my system. I'm not sure what to look for to see why it's not building here?

cormacpayne commented:
@KirkMunro Hey Kirk, would you mind pulling the latest changes from the `preview` branch? There were a couple of fixes we got out in the last day that affect the build.

KirkMunro commented:
@cormacpayne I pulled the latest changes and merged them into my fork. I also cleaned up the breaking change issue exceptions, because there were a _ton_ of duplicate entries, and many entries, duplicates or not, had paths that were specific to various users systems -- that's all cleaned up now. Lastly, I added some additional assembly version conflict exceptions that seemed to be required for a successful build (they weren't required for successful build the other day, but I couldn't get it to build without adding them now). It's building again locally so all _should_ be good with this commit.

markcowl commented:
@KirkMunro Thanks for the PR, and we appreciate the feedback.  The question of the verbs for these cmdlets is one we have often thought of.  Unfortunately, neither Connect / Disconnect nor Add/Remove are quite what non PowerShell gurus are used to for the simple act of logging in to Azure, so that the alias to login/logout is not likely to ever be removed.

Connect/Disconnect are likely closer to the actual action of the cmdlets, although because this is actually adding a set of contexts into the existing container, the current verbs are somewhat appropriate.  Given the longstanding usage of these cmdlet names, and the fact that the aliases will remain necessary, this doesn't seem to add enough value for us to accept it.

If you would like to discuss this further, we would be happy to do this.  Please feel free to email me (markcowl), or post to the MVP DL.

KirkMunro commented:
@markcowl Mark, if all you did was use a less appropriate verb like Add-/Remove- for connect/disconnect scenarios, but documented and presented proper cmdlet use and didn't encourage the use of aliases and going your own way when it comes to PowerShell, I wouldn't have such a problem with this (although in that case I would still have pushed the rename and then added aliases so that the old names could be used but the new names would enable discovery with the rest of the PowerShell ecosystem that is far larger than Azure and that understands that Connect-/Disconnect- are the right verbs to use here).

Instead, you/your team has chosen inappropriate verbs and then tried to compensate for it by creating aliases that define non-approved verbs. The argument that Login-/Logout- should be there for non-PowerShell gurus is really thin. It doesn't take much for someone to learn that Connect-/Disconnect- are the right verbs to use for such things, and by doing so, they will be able to make educated guesses and be right most of the time because the entire purpose of a limited set of verbs is to maintain consistency and discoverability. Except when you have teams _within Microsoft_ who decide to actively plant the seed through documentation and demos that it's ok to use whatever verbs you want in PowerShell via aliases. You might feel Login-/Logout- are more appropriate, and someone else may feel whatever verb they want is more appropriate than another approved verb, and so on, until different teams **from Microsoft that should be supporting and promoting the core tenets of PowerShell like consistency and discoverability** have widdled away at the notion of _using approved verbs and maintaining consistency and discoverability across all PowerShell modules and commands_ that has been there since day 1 when PowerShell shipped over 10 years ago.

I really don't understand you guys. The AzureRM cmdlets are very well known for doing so many things wrong in the PowerShell community. You are, however, open source, which allows community members to contribute to fixing some of those wrong doings. Yet you summarily reject this change, even though it is absolutely comprehensive and includes all necessary changes across the entire set of AzureRM modules (every error message, every command invoked) and across all open source Azure documentation (via the other PR). _All of it_.

You would do far better by dropping the current use of Login-/Logout-. Keep the aliases around for compatibility if you want, but they shouldn't be used by Microsoft in docs or in demos -- they may very well be the very first command that someone need to use in PowerShell after all if they're coming to it for the first time with a need to automate something with Azure.

Finally, for newcomers, you're doing them a far bigger favor by promoting proper, approved verbs that allow them to go further and discover other similar verbs (like Connect-AzureAD and Disconnect-AzureAD -- hello, does this make more sense now???) than by trying to show them verbs that may be used in a module here or there but that cannot be applied broadly across PowerShell as a whole.

markcowl commented:
@KirkMunro Again, thanks for the great feedback.  

I hope that having this PR closed does not dissuade you from submitting  further PRs in the future.  We were quite happy to see and accept your other submitted PR improving the visibility of our powershell-specific xml files, and hope you will submit more.  If we have dissuaded you from further PRs by not accepting this one, it will be our loss. By including the invitation to discuss this with the team directly, I was hoping to open a dialog about this.

We take conformance to PowerShell standards very seriously, we do see the significant benefit in being able to discover correct usage and apply it elsewhere, and this has been a major focus of our effort over the last few releases, so that usage of appropriate approved verbs, improved help validation, explicitly typed cmdlet output, support for proper confirmation, ability to use cmdlets in Jobs, conformance to parameter naming conventions, simplification of parameters, and the like are significantly improved in new cmdlets. We have instituted a strict review process for all new cmdlets with the guidance and participation of the PowerShell core team, and we think thsi has borne fruit in increased usability of new cmdlets.  We would really like your feedback on this.

As we apply these standards to existing cmdlets, we run into conflicting principles - we want to apply the standards to existing cmdlets without breaking customers, so that this happens slower than we would like in a perfect world.  You should see much more improvement over these areas in the next few releases.

In this particular case, usage of the possibly better verbs 'Connect/Disconnect' over 'Add/Remove' are outweighed in our estimation by longstanding usage and the desire to not introduce breaking changes unless absolutely necessary.  For reference, here is the actual documentation for each of these approved verb pairs.

https://msdn.microsoft.com/en-us/library/system.management.automation.verbscommon.add(v=vs.85).aspx

https://msdn.microsoft.com/en-us/library/system.management.automation.verbscommunications.connect(v=vs.85).aspx

I agree that by usage in some other sets of cmdlets, connect fits slightly better than Add, however, neither fits completely.  in this case, the action of the AzureAD cmdlets and ours are not identical, we are actually adding information into a local resource as a result of this cmdlet, and not simply connnecting a source to a target.

As far as discoverability and the use of login and logout aliases go, I am not casually speculating about customer experience,  I am speaking from direct user feedback and customer data.  Acquisition experiences are taken very seriously, and any barriers to customers becoming immediately productive with cmdlets receives a great deal of scrutiny.

I hope that we have not dissuaded you from continuing the great feedback and the contributions.  Please feel free to rech out over email, issues, or any other forum if you have fuirther feedback or questions.

danSpotter commented:
Thanks for this. I've now decided to name all my Get cmdlets to Acquire and all my Set cmdlets to Adjust.

mattmcnabb commented:
I agree with Kkrk here - the justification for using nonstandard verbs is virtually a non-argument. If someone who is not familiar with Powershell then the choice of which verb to use is arbitrary - they will have to learn a cmdlet name regardless.

KirkMunro commented:
@markcowl Thanks for the additional detail. I still disagree with the decision.

If you look at the Azure documentation from docs.microsoft.com that is open sourced in https://github.com/MicrosoftDocs/azure-docs, there are 245 references to Login-AzureRmAccount, compared to only 47 references to Add-AzureRmAccount. That shows that Microsoft is using Login-AzureRmAccount 5 times more in reference materials than they are using Add-AzureRmAccount. Is that by accident? Or is that because Add isn't really a very intuitive or appropriate verb to use for this command.

If I had simply added Connect-AzureRmAccount as an alias for Add-AzureRmAccount, would that have been accepted? I suspect it would, but if so, wouldn't it make sense to evaluate which is the better verb, Connect or Add? If a command with the verb Connect caches connection information (or adds context into an existing container) behind the scenes, is that really an argument for the command to use the verb Add instead? The command name should be as self explanatory as possible. Connect-AzureRmAccount says a lot more about what the command does to someone who has never used it before than Add-AzureRmAccount.

Looking just at the modules I have on my system, I have 31 commands that use the Connect verb, only 2 of which are aliases for other Connect verb commands, and none of which have Login aliases. And not surprisingly, they all seem to be used to connect to some server or service. In contrast, I only have 3 commands that use the Login verb, all of which come from AzureRM, and two of which are for the same command. In 10 years of using PowerShell to manage many, many different servers and services, I cannot think of one case where Add was used in place of Connect for a command used to connect (or login) to a server/service.

It is easy to understand why someone would go looking for a connect verb when they try to figure out how to connect to a service. I did when I first started using AzureRM, and I naturally want to type Connect every time I have to force myself to remember that I need to use Add instead for this and only this set of modules -- Every. Time. If I'm demonstrating this functionality to others, I have to explain around the awkwardness of this design, and tell them it should be called Connect-AzureRmAccount, but the AzureRM modules have a design flaw.

Regarding breaking changes, while I do feel that the Login- aliases should ideally go away over time (better for the community so that they converge on the use of Connect as the verb for such things), that would introduce breaking changes, and I understand you not wanting to introduce such changes. The good news is, you don't have to. You can keep Login-AzureRmAccount, Login-AzAccount, Logout-AzureRmAccount, Logout-AzAccount, Add-AzureRmAccount, and Remove-AzureRmAccount around forever...as aliases. That is what I set up in this PR. There are no breaking changes here. But there is a point that use of Login as a verb in documentation and in presentations should be discouraged. In the presentations I have seen, in the documentation I have read (and corrected in my other PR), Login is used 5 times more than the actual cmdlet verb. 5 times.

All of this speaks exactly to why this should change, without breaking a thing, by accepting this, and the related MicrosoftDocs/azure-docs PR.

brwilkinson commented:
+1 for Kirk's change. 

Connecting to azure, for  the majority of people is not an automated workflow anyway, due to the fact that people have to type in credentials and accept mfa. Which I think makes this change less impacful in breaking automated process. 

Hopefully this change does get approved. With these kinds of things, sooner is better.

gerane commented:
I do not understand this decision. There aren't any breaking changes and the use of add- makes absolutely no sense. I feel Kirk went above and beyond on this and I just do not understand the decision. I know I am not alone when I say that any module that uses add- to connect to a service, means I'm going to Google and having to consult documentation on how to connect. 

Verbs are for the user. They aren't to describe what the command is doing behind the scenes. Add a resource to a container makes no sense for a user. If we look at the verbs in the common contexts users are used to or will see the verbs, we can see why add- is a problematic.

Add-Computer vs Connect-Computer

Add-Computer - when are user sees this, it likely means they are adding a computer to a container of some sort, like a group, active directory, or similar. Nothing about this implies logging in or connecting. Add- has a more permanent feel to it, where connect implies more or what you'd expect when logging in or connecting to a service.

Connect-Computer - when a user sees this, it likely refers to connecting or interacting with a remote computer. This actually fits better than login, because connect implies connecting to a remote server or service. Login has always had more of a local feel in my opinion. 

Connect makes quick and easy since to all skill levels. If I look at a list of commands, and don't see login, connect will immediately stand out as making since. Add- would never cross my mind as being the command to connect to a service.

JPRuskin commented:
I agree with the arguments above.

As far as I can see, Connect is a far better verb than Add (especially by the definitions @markcowl linked). It's also not a breaking change, thanks to all the work put in by @KirkMunro in this PR.

I don't quite understand how best practice internally seems to be the use of a non-acceptable verb, and given the improvement on offer... rejecting it _before_ the offered discussion seems sad - setting the expectation, almost.

I would also support `Connect` being the default command name (rather than additional alias) as it has an acceptable verb, it's more understandable than `Add` for newcomers to PS standards (discoverability shouldn't be discounted), and it fits the suggested usage of the verb as defined by MS.

sdecker commented:
Another voice to say that closing and refuting the validity of this PR is wrong. It shows a significant amount of arrogance to ignore established rules on verbs and to push unapproved verbs and dig harder into a poor rebuttal. There is no breaking change the old command work so all scripts will still work.  Add makes ZERO sense. I am not adding an account. As for customer satisfaction and adoption, I absolutely struggled in the beginning to connect as the current command is nonsensical.  @markcowl, hi I'm a customer spending hundreds of thousands on Azure (and adding more) Login and Add are wrong!!! I would argue having "Account" is part of the problem. The command should be Connect-AzureRM. That is it because that is what I'm doing. I'm connecting so I can then do things other things.

If you want further proof, look at the Azure AD module. Guess what is has. "Connect-AzureAD"! Do they have different customers than you? Why are their customers not confused? Are they having barriers to use?

mattmcnabb commented:
It is worth mentioning that standard verbs are part of Microsoft's [Required development guidelines](https://msdn.microsoft.com/en-us/library/dd878238(v=vs.85).aspx#RD01), not the Strongly encouraged development guidelines. If developers are to be expected to follow these guidelines then Microsoft should lead by example.

markcowl commented:
@mattmcnabb @sdecker If you read the development guideline you cited on standard verb usage, it is about the verbs in cmdlet attributes.In this case, both cmdlet verbs 'Add' and 'Connect' are approved verbs from the list.

I could not be in greater agreement that all cmdlets must follow the required giuidelines, I'm just not sure how it applies to the actual issue at question.

mattmcnabb commented:
I see your point that both are accepted verbs, but Add is clearly defined as synonymous with "Add to, append or attach," whereas Connect means "Associate subsequent activies with a resource." That's a pretty clear difference.

markcowl commented:
@KirkMunro re-opening this PR to get this change started.  We are coordinating the change with the docs team.  Because we want to rolll out the documentation change with the code change, will have to decide whtehr this ships in December, or early in January.

I can submit a PR to your branch to being this up to date with the latest changes.

cormacpayne commented:
@KirkMunro Hey Kirk, I opened a PR in your repository to pull the latest changes from `Azure/preview` into your branch and resolve the merge conflicts you're seeing: https://github.com/KirkMunro/azure-powershell/pull/1

Locally, I pulled in the latest changes from the release branch and resolved those merge conflicts, but I will wait to push that until we have merged it back into the `preview` branch today.

KirkMunro commented:
@cormacpayne Thanks, I merged all of your changes into my PR.

cormacpayne commented:
@KirkMunro Hey Kirk, last release [we added two tests](https://github.com/Azure/azure-powershell/blob/preview/src/ResourceManager/Profile/Commands.Profile.Test/LoginCmdletTests.cs#L356-L407) to the `Commands.Profile.Test` project that use the old `AddAzureRMAccountCommand` class, which has been renamed to `ConnectAzureRmAccountCommand` in your PR. Would you mind updating these two tests with the new name of the command?

KirkMunro commented:
@cormacpayne Done.

cormacpayne commented:
@azuresdkci test this please

KirkMunro commented:
@cormacpayne: This PR seems to have stalled. I'm curious, what's holding this back?

maddieclayton commented:
@azuresdkci Test this please.

maddieclayton commented:
Resolved all unneeded changes to tools, and checked over other changes.  This should be good to go for next release.

maddieclayton commented:
@KirkMunro I reverted the dll because this is done automatically by the release tooling - this is for our own personal tooling that checks for breaking changes and PowerShell standard violations. 

Looks like you're right about Select-AzureRmProfile (I assumed this got left out when adding in the new aliases).  I'll remove it from the psd1.

For the two .csv files you mentioned, these are also for our tooling that checks for breaking changes.  I reverted the changes because the merge conflicts had become very complex and I wanted to ensure nothing was suppressed that had not been previously approved by our team.  I agree that at some point we should do cleanup around these files, but this is not a priority for our team as the StaticAnalysis will never touch the end user, and also will not affect any of our partners/developers.

maddieclayton commented:
@azuresdkci Test this please

markcowl commented on src/ResourceManager/Common/Commands.ResourceManager.Common/AzureRMCmdlet.cs:
```diff
@@ -102,7 +102,7 @@ protected override IAzureContext DefaultContext
                 if (DefaultProfile == null || DefaultProfile.DefaultContext == null || DefaultProfile.DefaultContext.Account == null)
                 {
-                    throw new PSInvalidOperationException("Run Login-AzureRmAccount to login.");
+                    throw new PSInvalidOperationException("Run Connect-AzureRmAccount to login.");
```
Please move this into a resource string

markcowl replied:
Or, better yet, use one of the resource strings below

maddieclayton replied:
Done.  Neither of the resource strings below directly apply so created a new one.

markcowl commented on src/Common/Commands.Common.Authentication/Properties/Resources.resx:
```diff
@@ -142,7 +142,7 @@
     <value>The environment name '{0}' is not found.</value>
   </data>
```
Changes to common code should include an addition to the ChangelOg for the Profile module

maddieclayton replied:
done

markcowl commented on src/ResourceManager/Compute/Commands.Compute/Extension/AzureDiskEncryption/Scripts/AzureDiskEncryptionPreRequisiteSetup.ps1:
```diff
@@ -72,7 +72,7 @@ $ErrorActionPreference = "Stop"
         if(-not $SvcPrincipals)
```
Each change to a module need to be accompanied by a change in the corresponding local changelog, to trigger our update and release of the module.

maddieclayton replied:
done

markcowl commented on src/ResourceManager/Profile/ChangeLog.md:
```diff
@@ -18,6 +18,8 @@
         - Additional information about change #1
 -->
 ## Current Release
+* `Add-AzureRmAccount` has been renamed as `Connect-AzureRmAccount`; an alias has been added for the old cmdlet name, and other aliases (`Login-AzAccount` and `Login-AzureRmAccount`) have been redirected to the new cmdlet name.
```
This should cover the common code changes, but we need a comment in each of the individual modules where we changed code

maddieclayton replied:
done

markcowl commented on src/ResourceManager/Profile/Commands.Profile/Context/GetAzureRMContext.cs:
```diff
@@ -55,7 +55,7 @@ public override void ExecuteCmdlet()
             {
                 WriteError(new ErrorRecord(
-                        new PSInvalidOperationException("Run Login-AzureRmAccount to login."),
+                        new PSInvalidOperationException("Run Connect-AzureRmAccount to login."),
```
Please move this into a resource string

maddieclayton replied:
done

markcowl commented on src/ResourceManager/Profile/Commands.Profile/Context/ImportAzureRMContext.cs:
```diff
@@ -27,6 +27,7 @@ namespace Microsoft.Azure.Commands.Profile
     /// Selects Microsoft Azure profile.
     /// </summary>
     [Cmdlet(VerbsData.Import, "AzureRmContext", SupportsShouldProcess = true, DefaultParameterSetName = ProfileFromDiskParameterSet), OutputType(typeof(PSAzureProfile))]
+    [Alias("Select-AzureRmProfile")]
```
This alias is confusing, please remove

maddieclayton replied:
done

markcowl commented on src/ResourceManager/Profile/Commands.Profile/help/Disconnect-AzureRmAccount.md:
```diff
@@ -5,67 +5,67 @@ online version: https://docs.microsoft.com/en-us/powershell/module/azurerm.profi
 
 ## SYNOPSIS
-Remove all credentials and contexts associated with the given account.
+Disconnect from a connected Azure account and removes all credentials and contexts associated with that account.
```
Disconnects a connected Azure account ...

maddieclayton replied:
done

markcowl commented on src/ResourceManager/Profile/Commands.Profile/help/Disconnect-AzureRmAccount.md:
```diff
@@ -5,67 +5,67 @@ online version: https://docs.microsoft.com/en-us/powershell/module/azurerm.profi
 ## DESCRIPTION
-Remove all credentials, and contexts (subscription and tenant information) associated with the given account.  After 
-executing this cmdlet, you will need to login again using Add-AzureRmAccount
+The Disconnect-AzureRmAccount cmdlet disconnects from a connected Azure account and removes all credentials and contexts (subscription and tenant information) associated with that account.
```
disconnects a ...

maddieclayton replied:
done

maddieclayton commented:
@azuresdkci Test this please

maddieclayton commented:
@azuresdkci Test this please

markcowl commented:
on demand run here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/385/

markcowl commented:
@maddieclayton Merge conflicts - please update and we should be good to go.
